ouroboros-praos/src/consensus/mod.rs

```diff
@@ -0,0 +1,349 @@
+
+        trace!("certified_leader_vrf: {}", certified_leader_vrf);
+        trace!("denominator: {}", denominator);
+        trace!("recip_q: {}", recip_q);
+        trace!("c: {}", self.c);
+        trace!("x: {}", x);
+
+        let ordering = x.exp_cmp(1000, 3, &recip_q);
+        match ordering.estimation {
+            ExpOrdering::LT => {
+                trace!(
+                    "Slot: {} - IS Leader: {} < {}",
+                    absolute_slot,
+                    recip_q,
+                    ordering.approx
+                );
+                true
+            }
+            _ => {
+                trace!(
+                    "Slot: {} - NOT Leader: {} >= {}",
+                    absolute_slot,
+                    recip_q,
+                    ordering.approx
+                );
+                false
+            }
+        }
+    }
+
+    /// Validate that the VRF key hash in the block matches the VRF key hash in the ledger
+    fn ledger_matches_block_vrf_key_hash(&self, pool_id: &PoolId, vrf_vkey: &[u8]) -> bool {
+        let vrf_vkey_hash = Hasher::<256>::hash(vrf_vkey);
+        trace!("block vrf_vkey_hash: {}", hex::encode(vrf_vkey_hash));
+        let ledger_vrf_vkey_hash = match self.pool_info.vrf_vkey_hash(pool_id) {
+            Ok(ledger_vrf_vkey_hash) => ledger_vrf_vkey_hash,
+            Err(error) => {
+                warn!("{:?} - {:?}", error, pool_id);
+                return false;
+            }
+        };
+        if vrf_vkey_hash.as_slice() != ledger_vrf_vkey_hash {
+            error!(
+                "VRF vkey hash in block ({}) does not match registered ledger vrf vkey hash ({})",
+                hex::encode(vrf_vkey_hash),
+                hex::encode(ledger_vrf_vkey_hash)
+            );
+            return false;
+        }
+        true
+    }
+
+    fn mk_vrf_input(&self, absolute_slot: u64, eta0: &[u8]) -> Hash<32> {
+        trace!("mk_vrf_input() absolute_slot {}", absolute_slot);
+        let mut hasher = Hasher::<256>::new();
+        hasher.input(&absolute_slot.to_be_bytes());
+        hasher.input(eta0);
+        hasher.finalize()
+    }
+}
+
+impl Validator for BlockValidator<'_> {
+    fn validate(&self) -> bool {
+        self.validate_babbage_compatible()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::consensus::BlockValidator;
+    use ctor::ctor;
+    use mockall::predicate::eq;
+    use ouroboros::ledger::{MockPoolInfo, PoolId, PoolSigma};
+    use ouroboros::validator::Validator;
+    use pallas_crypto::hash::Hash;
+    use pallas_math::math::{FixedDecimal, FixedPrecision};
+    use pallas_traverse::MultiEraHeader;
+    use std::fs::File;
+    use std::io::Read;
+
+    #[ctor]
+    fn init() {
+        // set rust log level to TRACE
+        // std::env::set_var("RUST_LOG", "ouroboros-praos=trace");
+
+        // initialize tracing crate
+        tracing_subscriber::fmt::init();
+    }
+
+    #[test]
+    fn test_validate_conway_block() {
```

Comment:
I feel like this is the kind of test that is a good candidate for snapshot testing: https://insta.rs/. We've been using this extensively within Aiken and it's quite enjoyable for capturing regressions and asserting against specific scenarios.

Not saying this should change _now_, but I might come and set that up eventually :)

Reply:
I've implemented snapshot tests. Please re-check.